File: api/src/main/java/io/github/runtimemodels/chazm/relation/manager/PossessesManager.java
package io.github.runtimemodels.chazm.relation.manager;

import io.github.runtimemodels.chazm.entity.Agent;
import io.github.runtimemodels.chazm.entity.Capability;
import io.github.runtimemodels.chazm.id.UniqueId;
import io.github.runtimemodels.chazm.relation.Possesses;

import java.util.Set;

/**
 * The {@linkplain PossessesManager} interface defines the APIs for managing {@linkplain Possesses}.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface PossessesManager {
	/**
	 * Creates a possesses relation between an {@linkplain Agent} and a {@linkplain Capability} in this {@linkplain PossessesManager}.
	 *
	 * @param agentId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @param capabilityId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Capability}.
	 * @param score
	 *            the <code>double</code> score of the possesses relation.
	 */
	void addPossesses(UniqueId<Agent> agentId, UniqueId<Capability> capabilityId, double score);

	/**
	 * Returns a set of {@linkplain Capability}s that is possessed by an {@linkplain Agent} in this {@linkplain PossessesManager}.
	 *
	 * @param id
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @return a set of {@linkplain Capability}s.
	 */
	Set<Capability> getPossesses(UniqueId<Agent> id);

	/**
	 * Returns a set of {@linkplain Agent}s that possesses a {@linkplain Capability} in this {@linkplain PossessesManager}.
	 *
	 * @param id
	 *            the {@linkplain UniqueId} that represents the {@linkplain Capability}.
	 * @return a set of {@linkplain Agent}s.
	 */
	Set<Agent> getPossessedBy(UniqueId<Capability> id);

	/**
	 * Returns the <code>double</code> score of the possesses relation between an {@linkplain Agent} and a {@linkplain Capability} in this
	 * {@linkplain PossessesManager}.
	 *
	 * @param agentId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @param capabilityId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Capability}.
	 * @return the <code>double</code> score.
	 */
	double getPossessesScore(UniqueId<Agent> agentId, UniqueId<Capability> capabilityId);

	/**
	 * Sets a new <code>double</code> score for the possesses relation between an {@linkplain Agent} and a {@linkplain Capability} in this
	 * {@linkplain PossessesManager}.
	 *
	 * @param agentId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @param capabilityId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Capability}.
	 * @param score
	 *            the new <code>double</code> score.
	 */
	void setPossessesScore(UniqueId<Agent> agentId, UniqueId<Capability> capabilityId, double score);

	/**
	 * Removes a possesses relation between an {@linkplain Agent} and a {@linkplain Capability} from this {@linkplain PossessesManager}.
	 *
	 * @param agentId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Agent}.
	 * @param capabilityId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Capability}.
	 */
	void removePossesses(UniqueId<Agent> agentId, UniqueId<Capability> capabilityId);

	/**
	 * Removes all possesses relations from this {@linkplain PossessesManager}.
	 */
	void removeAllPossesses();
}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/relation/PossessesManager.kt
package runtimemodels.chazm.api.relation

import runtimemodels.chazm.api.entity.Agent
import runtimemodels.chazm.api.entity.AgentId
import runtimemodels.chazm.api.entity.Capability
import runtimemodels.chazm.api.entity.CapabilityId

/**
 * The [PossessesManager] interface defines the APIs for managing a set of [Possesses] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface PossessesManager {
    /**
     * Adds a [Possesses] relation between an [Agent] and a [Capability] to this [PossessesManager].
     *
     * @param possesses the [Possesses] relation to add.
     */
    fun add(possesses: Possesses)
    /**
     * Returns the [Possesses] relation between an [Agent] and an [Capability] from this [PossessesManager].
     *
     * @param agentId           the [AgentId] that represents the [Agent].
     * @param capabilityId the [CapabilityId] that represents the [Capability].
     * @return the [Possesses] relation if it exists, `null` otherwise.
     */
    operator fun get(agentId: AgentId, capabilityId: CapabilityId): Possesses?

    /**
     * Returns a [Map] of [Capability]s that an [Agent] possesses in this [PossessesManager].
     *
     * @param id the [AgentId] that represents the [Agent].
     * @return a [Map] of [Capability]s.
     */
    operator fun get(id: AgentId): Map<CapabilityId, Possesses>

    /**
     * Returns a [Map] of [Agent]s that possesses a [Capability] in this [PossessesManager].
     *
     * @param id the [CapabilityId] that represents the [Capability].
     * @return a [Map] of [Agent]s.
     */
    operator fun get(id: CapabilityId): Map<AgentId, Possesses>

    /**
     * Removes a [Possesses] relation between an [Agent] and a [Capability] from this [PossessesManager].
     *
     * @param agentId      the [AgentId] that represents the [Agent].
     * @param capabilityId the [CapabilityId] that represents the [Capability].
     * @return the [Possesses] relation that was removed, `null` otherwise.
     */
    fun remove(agentId: AgentId, capabilityId: CapabilityId): Possesses?

    /**
     * Removes all [Possesses] relations associated to an [Agent] from this [NeedsManager].
     *
     * @param id the [AgentId] that represents the [Agent].
     */
    fun remove(id: AgentId)

    /**
     * Removes all [Possesses] relations associated to a [Capability] from this [NeedsManager].
     *
     * @param id the [CapabilityId] that represents the [Capability].
     */
    fun remove(id: CapabilityId)
}
